Add `--file_write_strategy` to Bazel

fmeum opened the pull request:
Work towards #24808

RELNOTES[NEW]: If the new `--file_write_strategy` flag is set to `lazy`, files produced by file write actions (e.g. the ones produced by `ctx.actions.write` and `ctx.actions.expand_template`) are not written to disk with BwoB unless required by local execution or upon explicit request. The default is `eager`, which retains the current behavior.

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java:
```diff
@@ -1004,6 +1017,109 @@ public boolean wasModifiedSinceDigest(Path path) {
     }
   }
 
+  public static final class FileWriteOutputArtifactValue extends FileArtifactValue
```
I made `FileWriteOutputArtifactValue` private, but also added `writeTo` to `FileArtifactValue` with a default implementation in terms of `getInputStream`. Having it implement `StreamWriter` just like `VirtualActionInput` is just very convenient for the use sites. Let me know if you find this too intrusive and I can refactor this further.

tjgq replied:
This works, thanks.

tjgq commented on src/main/java/com/google/devtools/build/lib/analysis/actions/LocalTemplateExpansionStrategy.java:
```diff
@@ -44,18 +44,43 @@ public ImmutableList<SpawnResult> expandTemplate(
+                            templateMetadata.substitutions(),
+                            pathResolver)));
+              } catch (EvalException e) {
+                throw new IllegalStateException(
+                    "Template expansion is not deterministic, first succeeded and then failed with: "
+                        + e.getMessage(),
+                    e);
```
It looks like the `getExpandedTemplateUnsafe` implementation doesn't actually throw `EvalException` (and there are no google3 overrides). Can we just amend the method signature?

tjgq commented on src/main/java/com/google/devtools/build/lib/remote/AbstractActionInputPrefetcher.java:
```diff
@@ -562,33 +563,52 @@ private Completable downloadFileNoCheckRx(
+                }
+              });
+    } else {
+      download =
```
Let's add a `checkState(metadata.isRemote())` here.

tjgq commented on src/main/java/com/google/devtools/build/lib/remote/merkletree/DirectoryTreeBuilder.java:
```diff
@@ -166,6 +166,16 @@ private static int buildFromActionInputs(
                   "missing metadata for '%s'",
                   input.getExecPathString());
+          if (metadata.isInline()) {
+            boolean childAdded =
+                currDir.addChild(
+                    FileNode.createExecutable(
+                        path.getBaseName(),
+                        metadata,
+                        DigestUtil.buildDigest(metadata.getDigest(), metadata.getSize()),
+                        toolInputs.contains(path)));
+            return childAdded ? 1 : 0;
+          }
```
I'd rather inline this into the `REGULAR_FILE` case below.

tjgq commented on src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java:
Mind adding tests in `FilesystemValueChecker` (you can take the existing tests for remote metadata as inspiration)?

tjgq commented on src/test/java/com/google/devtools/build/lib/remote/BuildWithoutTheBytesIntegrationTestBase.java:
```diff
@@ -1055,6 +1097,83 @@ public void downloadToplevel_unresolvedSymlink() throws Exception {
   }
 
+  @Test
+  public void downloadMinimal_fileWrite(
```
Can you also parameterize these tests for `ctx.actions.write_file` and `ctx.actions.expand_template`? (It's fine to use a dummy template with no substitutions, I'm only concerned about covering the separate code path in `LocalTemplateExpansionStrategy`)

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java:
```diff
@@ -1004,6 +1043,105 @@ public boolean wasModifiedSinceDigest(Path path) {
+
+    @Override
+    public InputStream getInputStream() {
+      // TODO: Avoid materializing the full content in memory by using a variant of
+      //  Piped{Input,Output}Stream that works well with virtual threads.
```
@justinhorvitz Is this something that has come up for you before? It could be particularly useful for `VirtualActionInput`'s `StreamWriter` implementation in Bazel.

justinhorvitz replied:
Uploading virutal inputs for remote execution can cost a lot of memory, so internally we use a semaphore to limit the number of inflight bytes. For large inputs, we use a "split store" technique where fixed-length chunks of the file are uploaded in parallel and then spliced together by the remote execution service. This means that we need to materialize the full file bytes, so using a piped stream wouldn't be helpful unless we changed the upload technique.

tjgq reviewed (commented):
LGTM, but I still need to figure out how to handle the discrepancy with the identically named, but differently valued `--file_write_strategy` flag at Google, and I couldn't get to it this week - so next Monday it will be.

meteorcloudy commented:
Looks like this needs a rebase now 😄

fmeum committed:
Lazily write files with Build without the Bytes

Implements lazy file writes for actions executed under Build without the
Bytes, so that the contents of write actions (e.g. write_file,
expand_template) are not materialized on disk until they are actually
needed, mirroring the treatment of remotely-stored outputs.

Squashed from the 24808-file-write development branch.

fmeum committed:
simplify

fmeum committed:
Update BUILD
